### PR TITLE
fix: implement the Dream core API

### PR DESCRIPTION
**Pull request by favonia**

### CODE REVIEW

File: src/core/logger/DiagnosticEmitterSigs.ml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  module TryWith (H : Handler) :
+  sig
+    val try_with : (unit -> H.result) -> H.result
+  end
+
+  module Perform :
+  sig
+    val print : Diagnostic.t -> unit
```
Comment (TOTBWF) [inline src/core/logger/DiagnosticEmitterSigs.ml:35]: Could we rename this to `emit`? I feel like `print` makes it seem as if things will be printed to stdout/stderr, which might not be the case.

File: src/core/Logger.mli [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  val messagef : ?loc:Span.t -> ?additional_marks:Span.t list -> code:Code.t -> ('a, Format.formatter, unit, Diagnostic.t) format4 -> 'a
+
+  (** [kmessagef kont ~loc ~additional_marks ~code format ...] constructs a diagnostic and then apply [kont] to the resulting diagnostic. *)
+  val kmessagef : (Diagnostic.t -> 'b) -> ?loc:Span.t -> ?additional_marks:Span.t list -> code:Code.t -> ('a, Format.formatter, unit, 'b) format4 -> 'a
+
+  (** [tracef ~loc format ...] record a frame. *)
+  val tracef : ?loc:Span.t -> ('a, Format.formatter, unit, (unit -> 'b) -> 'b) format4 -> 'a
+
+  (** [append_marks msg marks] appends [marks] to the additional marks of [msg]. *)
+  val append_marks : Diagnostic.t -> Span.t list -> Diagnostic.t
```
Comment (TOTBWF) [inline src/core/Logger.mli:20]: Perhaps this should be called `with_marks`?

**Review by TOTBWF — COMMENTED** [review]
Looks good! I've got a couple of naming nits, but nothing major.
However, I do want to think about how we handle `severity` a bit more before merging this.

File: src/core/logger/DiagnosticEmitterSigs.ml (hunk 1 above)
Comment (favonia) [reply src/core/logger/DiagnosticEmitterSigs.ml]: Done!

File: src/core/Logger.mli (hunk 2 above)
Comment (favonia) [reply src/core/Logger.mli]: For this I'm less sure. I feel `with_marks marks msg` could be interpreted in at least three ways:

1. Replace the additional marks in `msg` with `marks`.
2. Prepend `marks` to the additional marks in `msg`.
3. Append `marks` to the additional marks in `msg`.

The current name `append_marks` made it clear that it's 3. Any suggestions to address the ambiguity?

Comment (TOTBWF) [reply src/core/Logger.mli]: Good point! In that case `append_marks` is fine. The reason I suggested the change was to make
```
messagef ~code:SomeCode "some error"
|> append_marks some_marks
|> Perform.fatal
```

a little nicer to read, but this is fine! (and we may as well just use the `additional_marks` argument anyways)

**Pull request merged**